File: Python/ASP for Beginners/Trees/Depth-First-Search/Leetcode-94.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/description/

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def inorderTraversalIterative(self, root):
        stack = []
        outputList = []
        curr = root
        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.left
            curr = stack.pop()
            outputList.append(curr.val)
            curr = curr.right
            
            
        return outputList

# stack: 1
# curr = 1
# outputList:  

    # recursive
    def inorderTraversal(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[int]
        """
        outputList = []

        def dfs(root):
            if not root:
                return None

            dfs(root.left)
            outputList.append(root.val)
            dfs(root.right)

        dfs(root)
        return outputList

File: ASP Begginers/Trees/DFS/Leetcode-94.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    vector<int> tree;

public:
    vector<int> inorderTraversal(TreeNode* root) {
        inorder(root);
        return tree;
    }

    void inorder(TreeNode* root){
        if(!root){
            return;
        }
        inorder(root->left);
        tree.push_back(root->val);
        inorder(root->right);
    }
};